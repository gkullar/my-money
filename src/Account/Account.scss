.account {
  .card {
    width: 320px;
    min-height: 240px;
  }

  .account-title {
    margin-top: 0;
    margin-bottom: 16px;
  }

  .account-detail {
    margin-bottom: 10px;

    &:first-of-type {
      margin-top: 20px;
    }
  }

  // @todo use dynamic values for margin calc or flex(change html)
  .spinner {
    position: absolute;
    top: 50%;
    left: 50%;
    margin-top: 20px;
    margin-left: -12px;
  }
}
