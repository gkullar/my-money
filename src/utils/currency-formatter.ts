export const toGBP = (value: number): string => 'Â£' + (value / 100).toFixed(2);
