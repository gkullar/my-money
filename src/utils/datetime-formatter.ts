export const toDateTime = (value: string): string =>
  new Date(value).toLocaleString();
