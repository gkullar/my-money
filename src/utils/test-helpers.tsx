import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import 'jest-styled-components';
import AppProviders from '../app-providers';

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AppProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
