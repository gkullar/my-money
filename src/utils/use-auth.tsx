import React, {
  createContext,
  useContext,
  useState,
  FunctionComponent
} from 'react';
import uuid from './uuid';

const accessTokenKey = `${process.env.REACT_APP_API_ACCESS_TOKEN_KEY}`;
const stateTokenKey = `${process.env.REACT_APP_API_STATE_TOKEN_KEY}`;
const clientId = `${process.env.REACT_APP_API_CLIENT_ID}`;
const clientSecret = `${process.env.REACT_APP_API_CLIENT_SECRET}`;
const redirectUri = 'http://localhost:3000'; // app not for production!

if (!localStorage.getItem(stateTokenKey))
  localStorage.setItem(stateTokenKey, uuid());

const stateToken = localStorage.getItem(stateTokenKey);

interface State {
  isAuthenticated: boolean;
  login: () => void;
  logout: () => void;
}

const authContext = createContext<State>({} as any);

const useAuth = () => useContext(authContext);

const AuthProvider: FunctionComponent<{}> = ({ children }) => {
  const auth = useAuthProvider();
  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
};

function useAuthProvider(): State {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(
    localStorage.getItem(accessTokenKey) ? true : false
  );
  const params = new URLSearchParams(window.location.search);

  if (!isAuthenticated && params.has('code') && params.has('state')) {
    if (params.get('state') !== stateToken)
      throw new Error('Error: state token not matched'); // @todo handle error in UI

    authenticate(params.get('code') as string);
  }

  async function authenticate(code: string) {
    const data = new FormData();
    data.append('grant_type', 'authorization_code');
    data.append('client_id', clientId);
    data.append('client_secret', clientSecret);
    data.append('redirect_uri', redirectUri);
    data.append('code', code);

    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/oauth2/token`,
      {
        method: 'post',
        body: data
      }
    );

    const result = await response.json();

    if (response.status !== 200 || result.error)
      throw new Error('Error: Failed Authentication'); // @todo handle error in UI

    localStorage.setItem(accessTokenKey, result.access_token);
    setIsAuthenticated(true);
    window.history.replaceState(null, document.title, window.location.pathname); // remove query params
  }

  function login() {
    window.location.href = `https://auth.monzo.com/?client_id=${clientId}&redirect_uri=${redirectUri}&response_type=code&state=${stateToken}`;
  }

  async function logout() {
    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/oauth2/logout`,
      {
        method: 'post',
        headers: new Headers({
          Authorization: `Bearer ${localStorage.getItem(accessTokenKey)}`
        })
      }
    );

    const result = await response.json();

    if (!result.ok) throw new Error('Error: Failed to logout'); // @todo handle error in UI

    localStorage.removeItem(accessTokenKey);
    setIsAuthenticated(false);
  }

  return { isAuthenticated, login, logout };
}

export { useAuth, AuthProvider };
