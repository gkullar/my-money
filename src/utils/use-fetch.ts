import { useState, useEffect } from 'react';

const accessTokenKey = `${process.env.REACT_APP_API_ACCESS_TOKEN_KEY}`;

interface State<T> {
  data: T;
  loading: boolean;
  hasErrors: boolean;
  hasPermissions: boolean;
}

const useFetch = <T>(url: string): State<T> => {
  const [data, setData] = useState<T>({} as T);
  const [loading, setLoading] = useState<boolean>(true);
  const [hasErrors, setHasErrors] = useState<boolean>(false);
  const [hasPermissions, setHasPermissions] = useState<boolean>(true);

  async function fetchUrl() {
    const response = await fetch(url, {
      method: 'get',
      headers: new Headers({
        Authorization: `Bearer ${localStorage.getItem(accessTokenKey)}`
      })
    });

    const result = await response.json();

    if (response.status === 401) {
      localStorage.removeItem(accessTokenKey);
      window.location.href = '/';
    } else if (response.status === 403) {
      setHasPermissions(false);
    } else if (response.status !== 200) {
      throw new Error('Error: Failed Data Fetch'); // @todo handle error in UI
    } else {
      setData(result);
      setLoading(false);
      setHasErrors(false);
      setHasPermissions(true);
    }
  }

  useEffect(() => {
    fetchUrl();
  }, []);

  return { data, loading, hasErrors, hasPermissions };
};

export default useFetch;
