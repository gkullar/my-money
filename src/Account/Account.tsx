import React from 'react';
import './Account.scss';
import AccountDetail from './AccountDetail/AccountDetail';
import useFetch from '../_hooks/useFetch';
import Card from '../Card/Card';
import Spinner from '../Spinner/Spinner';

interface Props {
  id: string;
  title: string;
}

interface State {
  balance: number;
  total_balance: number;
  currency: string;
  spend_today: number;
}

export default function Account(props: Props) {
  const { data, loading } = useFetch<State>(
    `${process.env.REACT_APP_API_URL}/balance?account_id=${props.id}`
  );

  return (
    <div className="account">
      <Card>
        <h2 className="account-title">{props.title}</h2>
        <hr />
        {loading ? (
          <Spinner />
        ) : (
          <div>
            <AccountDetail title="Balance" value={data.balance} />
            <AccountDetail
              title="Available Balance"
              value={data.total_balance}
            />
            <AccountDetail title="Spend Today" value={data.spend_today} />
          </div>
        )}
      </Card>
    </div>
  );
}
