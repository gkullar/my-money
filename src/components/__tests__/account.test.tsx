import React from 'react';
import Account from '../account';
import { render, wait } from '../../utils/test-helpers';

describe('Account', () => {
  test('should render an error message if error', async () => {
    const spy = jest.spyOn(window, 'fetch') as jest.Mock<any>;
    spy.mockResolvedValue({ status: 404 });

    const { container } = render(
      <Account id="TestId" accountNumber="TestAccountNo" />
    );

    await wait(() =>
      expect(container.textContent).toContain(
        'There was an error retrieving account data'
      )
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render account details if no error', async () => {
    const spy = jest.spyOn(window, 'fetch') as jest.Mock<any>;
    spy.mockResolvedValue({
      status: 200,
      json: () =>
        Promise.resolve({
          balance: 100,
          total_balance: 100,
          spend_today: 0
        })
    });

    const { container } = render(
      <Account id="TestId" accountNumber="TestAccountNo" />
    );

    await wait(() => expect(container.textContent).toContain('TestAccountNo'));

    expect(container.firstChild).toMatchSnapshot();
  });
});
