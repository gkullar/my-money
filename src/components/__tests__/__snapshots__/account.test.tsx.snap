// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Account should render account details if no error 1`] = `
.c1 {
  padding: 20px;
  position: relative;
  border-radius: 15px;
  color: #fff;
  background: rgba(0,219,163,0.9);
  box-shadow: 8px 8px 4px -4px rgba(95,107,125,0.4);
}

.c0 {
  width: 310px;
  height: 225px;
}

.c4 {
  font-size: 0.75rem;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  display: inline-block;
  margin-right: 15px;
}

.c7 {
  font-size: 0.75rem;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  margin-bottom: 5px;
}

.c5 {
  font-size: 1.3rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  display: inline-block;
}

.c8 {
  font-size: 1.3rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  font-weight: bold;
  font-size: 1.5rem;
}

.c11 {
  font-size: 1.3rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.c2 {
  padding: 0px 5px;
}

.c3 {
  margin-top: 5px;
  margin-bottom: 30px;
}

.c6 {
  border-bottom: 1px solid #fff;
  padding-bottom: 20px;
  margin-bottom: 20px;
}

.c9 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c10 {
  width: 50%;
}

@media (min-width:30em) {
  .c0 {
    width: 350px;
  }
}

<div
  class="c0 c1"
>
  <div
    class="c2"
  >
    <div
      class="c3"
    >
      <div
        class="account-detail"
      >
        <div
          class="c4"
        >
          Account No
        </div>
        <div
          class="c5"
        >
          TestAccountNo
        </div>
      </div>
    </div>
    <div
      class="c6"
    >
      <div
        class="account-detail"
      >
        <div
          class="c7"
        >
          Balance
        </div>
        <div
          class="c8"
        >
          £1.00
        </div>
      </div>
    </div>
    <div
      class="c9"
    >
      <div
        class="c10"
      >
        <div
          class="account-detail"
        >
          <div
            class="c7"
          >
            Available
          </div>
          <div
            class="c11"
          >
            £1.00
          </div>
        </div>
      </div>
      <div
        class="c10"
      >
        <div
          class="account-detail"
        >
          <div
            class="c7"
          >
            Todays Spend
          </div>
          <div
            class="c11"
          >
            £0.00
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Account should render an error message if error 1`] = `
.c1 {
  padding: 20px;
  position: relative;
  border-radius: 15px;
  color: #fff;
  background: rgba(0,219,163,0.9);
  box-shadow: 8px 8px 4px -4px rgba(95,107,125,0.4);
}

.c0 {
  width: 310px;
  height: 225px;
}

@media (min-width:30em) {
  .c0 {
    width: 350px;
  }
}

<div
  class="c0 c1"
>
  <p>
    There was an error retrieving account data
  </p>
</div>
`;
